plugins/inputs/eventhub/eventhub.go

```diff
@@ -0,0 +1,382 @@
+		receiveOpts = append(receiveOpts, eventhub.ReceiveWithPrefetchCount(e.PrefetchCount))
+	}
+
+	if e.Epoch != 0 {
+		receiveOpts = append(receiveOpts, eventhub.ReceiveWithEpoch(e.Epoch))
+	}
+
+	return receiveOpts, err
+}
+
+func (e *EventHub) onMessage(ctx context.Context, event *eventhub.Event) (err error) {
+
+	metrics, err := e.parser.Parse(event.Data)
+
+	if err != nil {
+		log.Printf("E! [inputs.eventhub] error %s", err)
+		return err
+	}
+
+	log.Printf("D! [inputs.eventhub] %d metrics found after parsing", len(metrics))
+
+	for i := range metrics {
+		metrics[i].AddField(e.SystemPropertiesPrefix+"SequenceNumber", *event.SystemProperties.SequenceNumber)
+
+		if e.EnqTimeTs {
+			metrics[i].SetTime(*event.SystemProperties.EnqueuedTime)
+		} else {
+			metrics[i].AddField(e.SystemPropertiesPrefix+"EnqueuedTime", (*event.SystemProperties.EnqueuedTime).UnixNano()/int64(time.Millisecond))
+		}
+
+		metrics[i].AddField(e.SystemPropertiesPrefix+"Offset", *event.SystemProperties.Offset)
+
+		if event.SystemProperties.PartitionID != nil {
+			metrics[i].AddTag(e.SystemPropertiesPrefix+"PartitionID", string(*event.SystemProperties.PartitionID))
+		}
+		if event.SystemProperties.PartitionKey != nil {
+			metrics[i].AddTag(e.SystemPropertiesPrefix+"PartitionKey", *event.SystemProperties.PartitionKey)
+		}
+		if event.SystemProperties.IoTHubDeviceConnectionID != nil {
+			metrics[i].AddTag(e.SystemPropertiesPrefix+"IoTHubDeviceConnectionID", *event.SystemProperties.IoTHubDeviceConnectionID)
+		}
+		if event.SystemProperties.IoTHubAuthGenerationID != nil {
+			metrics[i].AddTag(e.SystemPropertiesPrefix+"IoTHubAuthGenerationID", *event.SystemProperties.IoTHubAuthGenerationID)
+		}
+		if event.SystemProperties.IoTHubConnectionAuthMethod != nil {
+			metrics[i].AddTag(e.SystemPropertiesPrefix+"IoTHubConnectionAuthMethod", *event.SystemProperties.IoTHubConnectionAuthMethod)
+		}
+		if event.SystemProperties.IoTHubConnectionModuleID != nil {
+			metrics[i].AddTag(e.SystemPropertiesPrefix+"IoTHubConnectionModuleID", *event.SystemProperties.IoTHubConnectionModuleID)
+		}
+		if event.SystemProperties.IoTHubEnqueuedTime != nil {
+			if e.IotHubEnqTimeTs {
+				metrics[i].SetTime(*event.SystemProperties.IoTHubEnqueuedTime)
+			} else {
+				metrics[i].AddField(e.SystemPropertiesPrefix+"IoTHubEnqueuedTime", (*event.SystemProperties.IoTHubEnqueuedTime).UnixNano()/int64(time.Millisecond))
+			}
+		}
+	}
+
+	id := e.acc.AddTrackingMetricGroup(metrics)
+
+	e.tracker.mux.Lock()
+	e.tracker.messages[id] = metrics
+	e.tracker.mux.Unlock()
+
+	return nil
+}
+
+func (e *EventHub) startTracking(ctx context.Context) {
+	defer e.wg.Done()
+	for {
+		select {
+		case <-ctx.Done():
+			if len(e.tracker.messages) == 0 { // everything has been delivered
+				return
+			}
+		case DeliveryInfo := <-e.acc.Delivered():
+			log.Printf("D! [inputs.eventhub] tracking:: ID %d delivered: %t", DeliveryInfo.ID(), DeliveryInfo.Delivered())
+
+			if DeliveryInfo.Delivered() {
```

Comment:
I don't think consumption rate will be limited when an in memory persister is used? It should be possible to add a third type of persister to azure-event-hubs-go which is something like a batched file persister (write every N events instead of every single event). If I find the time I can create an issue. 

I think your other points on out of order processing and concurrency are perfectly valid, but do we need to include this functionality in the plugin? Or is expanding the azure-event-hubs-go repository the preferred way to do this? The latter option might be more beneficial to the broad public.

Reply:
There are a few things we generally want to be able to do in a plugin like this that consumes from a message queue:
- Take at least enough metrics to trigger an immediate flush in the output plugin, by filling the metric batch.
- Wait for these metrics to be written by the output plugin before taking more messages, we don't want to consume to far ahead.
- Depending on the queue, acknowledge or update offsets when a message is successfully written, and reject messages that are lost.  In Telegraf messages are lost if the metric buffer fills.

I looked into the azure-event-hubs-go code a bit, and it appears that the handler is called in the same goroutine for each message.  Once the handler returns, the message is immediately marked as acknowledged if there is no error and the offsets are persisted. If the handler returns an error the message is marked as delivery failed.

This means there is no way to take multiple messages and report back to Event Hub later on if they were processed successfully or not.

The only way it can work is for us to report that the message are handled successfully and hope it can be written by the output.

Since you are re-adding metrics if they weren't able to be delivered, the only way we could still lose metrics is on shutdown/restart/crash and only if the output is down.  It isn't a full durability guarantee but I think it is the best we can do with this library.

The other thing we need to do is to stop consuming messages once we have `max_undelivered_messages` in flight.  Otherwise we will just keep grabbing new messages even though nothing is being written.

Reply:
See the bottom of https://github.com/influxdata/telegraf/pull/6928#issuecomment-599225047.